KDCSP-49 Merge MongoDB Adapter with latest ADAGUC

mtsr opened the pull request:
Contains KDC-specific changes which may need to be refactored before upstream merge

Commit:
KDCSP-49 Merge MongoDB Adapter with latest ADAGUC
Contains KDC-specific changes which may need to be refactored before upstream merge

saskiawagenaar commented on adagucserverEC/CAutoResource.cpp:
```diff
@@ -3,6 +3,12 @@
 #include "CServerError.h"
 #include "CDFObjectStore.h"
 
+#ifdef ADAGUC_USE_KDCMONGODB
```
[KDC-Specific]
KDC moeten we uit de naam halen als we een algemene MongoDB adapter aan willen bieden.

saskiawagenaar commented on adagucserverEC/CAutoResource.cpp:
```diff
@@ -40,23 +46,44 @@ int CAutoResource::configureDataset(CServerParams *srvParam,bool plain){
+        return 1;
+      }
+    #endif
+    #ifdef ADAGUC_USE_KDCMONGODB
```
[KDC-Specific]
Als ik het goed begrijp is dit geen MongoDB specifieke wijziging, maar een wijziging waarbij we de dataset config uit de database halen i.p.v. uit de config file. Dit moeten we denk ik splitsen van de MongoDB implementatie.

mtsr replied:
Eens. Hetzelfde geldt dan voor het wegschrijven van de dataset config naar de database.

saskiawagenaar commented on adagucserverEC/CAutoResource.cpp:
```diff
@@ -40,23 +46,44 @@ int CAutoResource::configureDataset(CServerParams *srvParam,bool plain){
-      return 1;
-    }
+      CDBAdapterMongoDB *mongoDB = new CDBAdapterMongoDB();
+      if (mongoDB == NULL) {
```
Zouden we deze check ook op nog meer plekken toe moeten voegen?

mtsr replied:
Voor de mongoDBadapter een prima plan. Daarna misschien voorstellen om dat meer te doen in de ADAGUC code. Misschien ook iets waar sonarqube iets over vindt?

saskiawagenaar commented on adagucserverEC/CDBFactory.cpp:
```diff
@@ -32,6 +32,7 @@ const char *CDBFactory::className="CDBFactory";
 CDBAdapter *CDBFactory::staticCDBAdapter = NULL;
 
 CDBAdapter *CDBFactory::getDBAdapter(CServerConfig::XMLE_Configuration *cfg){
+  CDBDebug("getDBAdapter");
```
Is dit debug statement nuttig?

saskiawagenaar replied:
Op regel 57 zou ik het debug statement wel aanzetten, om het consistent te houden met de andere database types. Ik mag helaas op die plek geen comment plaatsen.

saskiawagenaar commented on adagucserverEC/CGetFileInfo.cpp:
```diff
@@ -8,7 +8,11 @@ CT::string CGetFileInfo::getLayersForFile(const char *filename){
 
   CT::string fileInfo="";
-       
+
```
Is dit een soort algemene bugfix uit de KDC Adaguc repository die nog niet doorgevoerd is op de algemene Adaguc repository?

mtsr replied:
@Flurb ?

saskiawagenaar commented on adagucserverEC/CRequest.cpp:
```diff
@@ -2174,7 +2174,13 @@ int CRequest::process_querystring(){
   
-  
+
+  #ifdef ADAGUC_USE_KDCMONGODB
```
[KDC-Specific]
Gek ook dat dit alleen voor Mongo nodig is.

mtsr replied:
Mogelijk kan setConfig een statische methode worden die de global *configurationObject aanpast. Dan kan de new () in elk geval weg.

Flurb replied:
De aanpassing in deze file is voor datasetvisualisatie.
Als je een dataset wil visualiseren, dan moet de config eigenlijk aangepast worden, namelijk de parameters van AutoResource (zoals hier beneden beschreven). Anders doet deze het niet.
Als je daarna een individuele wil visualiseren, moet de config weer omgezet worden.

saskiawagenaar commented on adagucserverEC/CRequest.cpp:
```diff
@@ -2487,6 +2493,10 @@ int CRequest::process_querystring(){
             CT::string *hashList=values[1].splitToArray("#");
             srvParam->autoResourceLocation.copy(hashList[0].c_str());
             delete[] hashList;
+            if(srvParam->cfg->AutoResource.size()>0){
```
Is dit een soort algemene bugfix die alleen op onze repository is doorgevoerd?

mtsr replied:
@Flurb ?

saskiawagenaar commented on adagucserverEC/CRequest.cpp:
```diff
@@ -2496,6 +2506,10 @@ int CRequest::process_querystring(){
         if(value0Cap.equals("DATASET")){
           if(srvParam->datasetLocation.empty()){
             srvParam->datasetLocation.copy(values[1].c_str());
+            if(srvParam->cfg->AutoResource.size()>0){
```
Is dit een soort algemene bugfix die alleen op onze repository is doorgevoerd?

mtsr replied:
@Flurb ?

saskiawagenaar commented on adagucserverEC/CServerConfig_CPPXSD.h:
```diff
@@ -738,6 +738,16 @@ class CServerConfig:public CXMLSerializerInterface{
           if(equals("value",5,attrname)){attr.value.copy(attrvalue);return;}
         }
     };
+    class XMLE_VFSPath: public CXMLObjectInterface{
```
[KDC-specific]
KDC specifieke wijziging.

saskiawagenaar commented on adagucserverEC/CServerParams.cpp:
```diff
@@ -543,10 +543,16 @@ bool CServerParams::checkTimeFormat(CT::string& timeToCheck){
 int CServerParams::parseConfigFile(CT::string &pszConfigFile){
   CT::string configFileData;
-  
+  CT::string subStringOfConfigString = pszConfigFile;
```
Hier zou van mij wel wat commentaar bij mogen.. 
Het lijkt mij dat dit ook een wijziging in het kader van het ophalen van de config file uit de database is.

saskiawagenaar commented on adagucserverEC/adagucserver.cpp:
```diff
@@ -252,6 +252,7 @@ int main(int argc, const char *argv[]){
         setDebugFunction(serverDebugFunction);
         
         CT::string fileInfo = CGetFileInfo::getLayersForFile(file.c_str());
+        CDFObjectStore::getCDFObjectStore()->clear();
```
Ik snap niet zo goed waarom we dit hier doen, is dit ook een bugfix die op de KDC respository is gedaan?

mtsr committed:
Aanpassing debug logging in getDBAdapter

Commit:
Automatically reformatted COpenDAPHandler.cpp

Commit:
Removed unused commented out code and added #ifdefs around debug statements which were commmented out.

Commit:
Removed duplicate code which was not reachable.

Commit:
Refactored a method to obtain request information from the path such that the method HandleOpenDAPRequest becomes smaller.

Commit:
Refactored a method which handles autoconfiguration such that the method HandleOpenDAPRequest becomes smaller.

Commit:
Refactored a method to initialize the layer such that the method HandleOpenDAPRequest becomes smaller.

Commit:
Made it possible to control the order in which variables are written to a NetCDF file. One option is to write the dimension variables first and then the other variables (previously the default option), the other option is to keep the original order. This is necessary when we use the cache to read/write header files, but we do not use the cache to retrieve the data.

Commit:
First version of updated implementation of getFilesForIndices. We now select all granules and then loop over the dimension time until we reached the required start position.

Commit:
Return in one statement now instead of multiple.

Commit:
Added a space (' ') as a valid token for OPeNDAP urls, such that it is also possible to use OPeNDAP on datasets which have a space in their name.

Commit:
[KDCSP-180] Als gebruiker wil ik performance winst in OPeNDAP krijgen door het bestandspad van de header file uit de database te halen

Functie toegevoegd aan de MongoDB Adapter waarmee 1 pad van 1 granule uit een geselecteerde dataset kan worden gehaald.
Wordt gebruikt in elk OPeNDAP request om een header op te halen.

Commit:
[KDCSP-180] Als gebruiker wil ik performance winst in OPeNDAP krijgen door het bestandspad van de header file uit de database te halen

Gebruik van #ifdef in gebruik genomen in de OPeNDAP handler CPP

Commit:
[KDCSP-183] Splitted the method getFilesForIndices into two parts.

Commit:
[KDCSP-180] Als gebruiker wil ik performance winst in OPeNDAP krijgen door het bestandspad van de header file uit de database te halen

Known bug fixed ( Datasets with spaces in technical name are now allowed )

Commit:
[KDCSP-182] ADAGUC config "Cache" type aanpasbaar maken
Cache type is nu aanpasbaar in de configfile. Opties zijn header, all en none.

Commit:
[KDCSP-182] ADAGUC config "Cache" type aanpasbaar maken
Editted some comments about types of config that can be used.

Commit:
[KDCSP-182] ADAGUC config "Cache" type aanpasbaar maken
Bug gefixed; ADAGUC zonder cache doet het weer.

Commit:
[KDCSP-181] Removed debug statements which referred to variables which do not exist anymore.

Commit:
[KDCSP-192] First change: The header file which is opened in OPeNDAP is no longer stored in the CCDFObjectStore. As a result, this header file will not be cleared when the CCDFObjectStore is cleared. Second change: The variable time will now always be broadcasted as a double in the header of an OPeNDAP request, and it will always be written as a double. As a result, we can now always use the value from the database instead of depending on the type of the original time value.

Commit:
[KDCSP-191] - Added some documentation to the openCDFObjectHeader method.

Commit:
[KDCSP-191] - Added a comment explaining why we do not store the header in the store.

Commit:
[KDCSP-191] - Clarified in the name that this method does NOT add the object to the store and does not use the store to access it.

Commit:
[KDCSP-183] - Updated the different names of Mongo query results to one standard name: 'queryResultCursor'.

Commit:
[KDCSP-182] ADAGUC config "Cache" type aanpasbaar maken
Fixes after review. Using a CacheType enum now.

Commit:
[KDCSP-183] - Named 'magical' constants nToSkip and nToReturn.

Commit:
[KDCSP-182] ADAGUC config "Cache" type aanpasbaar maken
Little bug fixed. Cache is only enabled when "header" or "data" is selected.

Commit:
[KDCSP-183] - Directly jump to the correct step within a granule.

Commit:
Merge branch 'KDCSP-131' of https://github.com/mtsr/adaguc-server into KDCSP-131

Commit:
[KDCSP-180] - Merged KDCSP-180 into KDCSP-131.

Commit:
[KDCSP-180] Als gebruiker wil ik performance winst in OPeNDAP krijgen door het bestandspad van de header file uit de database te halen
Check toegevoegd dat er alleen gezocht wordt naar granules waar het "adaguc" veld bestaat.

Commit:
[KDCSP-180] Als gebruiker wil ik performance winst in OPeNDAP krijgen door het bestandspad van de header file uit de database te halen
Bug gefixed waarbij filepaden half werden weggeschreven in de cache.

Commit:
[KDCSP-191] - Fixed problem where header file was opened multiple times.

Commit:
[KDCSP-182] - Fixed problem where strings were not handled correctly in de mongo adapter.

saskiawagenaar committed:
[KDCSP-216] - Fixed issue where preview was not working due to the configuration file which was not passed to the mongo adapter correctly and threadsafe in all cases.

saskiawagenaar committed:
[KDCSP-227] - Fixed issue in mongo adapter for datasets without the dimension time. For these datasets, exceptions would occur in the mongo adapter when checking if the dataset is aggregated.

saskiawagenaar committed:
[KDCSP-227] - Fixed issue where strings were not passed along correctly. Maarten pointed out that CT::string is the preferred class to pass Strings around. Also fixed a bug where not all opendap records were recognized for aggregated sets where not all granules have the same number of timesteps.

Commit:
Fix for datasets without a dimension

Commit:
Added multi dimension support for granules in MongoDB

Commit:
Dimension values are now stored in 'adaguc.dimensions' object AND outside the 'adaguc.dimensions' object

Commit:
CDBAdapterMongoDB.cpp now reads from database FIRST in adaguc.dimensions, then adaguc for dimensions

Commit:
Fixed bug in GetFilesForIndicesAndDimensions; using old or new dimensions object now depends on query

Commit:
Removed unnecessary logging

Commit:
Bugfix for OPeNDAP(Cannot sort with arrays)

Commit:
Fixes after review

Commit:
Added some comments to new code (code wasn't really clear)

Commit:
Fixed a bug where dimension values were stored with 1 character deficit

saskiawagenaar committed:
[KDCSP-335] - Dataset name and version are now separated by :: instead of _ for opendap.

saskiawagenaar committed:
[KDCSP-335] - Updated OPeNDAP ::-separator regex, made it more explicit.

neutvd committed:
[EUT-185] Fixes OPeNDAP compatibility with Java NetCDF library.

See also
https://github.com/KNMI/adaguc-server/commit/75109dfc638fa1cab820ec6c00f6fcf2e2aaa907
for the same fix in the main repository.
